#!/bin/bash
threads=6
inputs_dir="/home/bioinformatikai/HW2/inputs"
outputs_dir="/home/bioinformatikai/HW2/outputs"

prefetch -O /home/bioinformatikai/HW2/inputs ERR204044 SRR15131330 SRR18214264
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/022/832/545/GCF_022832545.1_ASM2283254v1/GCF_022832545.1_ASM2283254v1_genomic.fna.gz
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/022/832/545/GCF_022832545.1_ASM2283254v1/GCF_022832545.1_ASM2283254v1_protein.faa.gz

fastq-dump --outdir /home/bioinformatikai/HW2/inputs/ --gzip --split-files /home/bioinformatikai/HW2/inputs/ERR204044
fastq-dump --outdir /home/bioinformatikai/HW2/inputs/ --gzip --split-files /home/bioinformatikai/HW2/inputs/SRR15131330
fastq-dump --outdir /home/bioinformatikai/HW2/inputs/ --gzip --split-files /home/bioinformatikai/HW2/inputs/SRR18214264

# Run FASTQC 
mkdir -p ${outputs_dir}/raw_data
fastqc -t ${threads} ${inputs_dir}/*fastq.gz -o ${outputs_dir}/raw_data

#The overall quality of the sequences does not look too bad. 
However, the sequence SRR18214264 reaches the red zone when measuring per-base sequence quality. 
There are significant levels of duplication in sequences

# Run standard FASTQ trimming and re-run FASTQC on cleaned FASTQ files
for i in ${inputs_dir}/*_1.fastq.gz;
do
  R1=${i};
  R2="${inputs_dir}/"$(basename ${i} _1.fastq.gz)"_2.fastq.gz";
  trim_galore -j ${threads} -o ${outputs_dir} --paired --length 20 ${R1} ${R2}
done

fastqc -t ${threads} ${outputs_dir}/*fq.gz -o ${outputs_dir}

#After rerunning FastQC, there were not many significant changes, however, the quality of reads slightly improved.
#The ERR204044_2_val_2 sample was not fully trimmed, however, that should not cause any significant problems.
#The per-base sequence quality of SRR18214264 was improved

# Create MultiQC plots for raw and processed data
multiqc -o ${outputs_dir}/raw_data ${outputs_dir}/raw_data
multiqc -o ${outputs_dir} ${outputs_dir}


##############################################################################################################################

#Assemble genomes using spades program.
spades.py --pe1-1 ERR204044_1_val_1.fq --pe1-2 ERR204044_2_val_2.fq -o /home/bioinformatikai/HW2/outputs/spades/ERR204044
spades.py --pe1-1 SRR15131330_1_val_1.fq --pe1-2 SRR15131330_2_val_2.fq -o /home/bioinformatikai/HW2/outputs/spades/SRR15131330
spades.py --pe1-1 SRR18214264_1_val_1.fq --pe1-2 SRR18214264_2_val_2.fq -o /home/bioinformatikai/HW2/outputs/spades/SRR18214264

#Assemble genomes using abyss program.
abyss-pe k=64 name=ERR204044 -C /home/bioinformatikai/HW2/outputs/abyss/ERR204044 in='/home/bioinformatikai/HW2/outputs/ERR204044_1_val_1.fq /home/bioinformatikai/HW2/outputs/ERR204044_2_val_2.fq'
abyss-pe k=64 name=SRR15131330 -C /home/bioinformatikai/HW2/outputs/abyss/SRR15131330 in='/home/bioinformatikai/HW2/outputs/SRR15131330_1_val_1.fq /home/bioinformatikai/HW2/outputs/SRR15131330_2_val_2.fq'
abyss-pe k=64 name=SRR18214264 -C /home/bioinformatikai/HW2/outputs/abyss/SRR18214264 in='/home/bioinformatikai/HW2/outputs/SRR18214264_1_val_1.fq /home/bioinformatikai/HW2/outputs/SRR18214264_2_val_2.fq'

#While there are slight differences in the specific metrics, 
#the overall trends and characteristics of the assemblies are comparable.
#The assemblies generated by different tools exhibit similar levels of genome coverage, duplication, and contig lengths
#In terms of contig lengths, the largest alignments range from approximately 41271 to 51210 base pairs, 
#and the NGA50 values range from approximately 5135 to 8874 base pairs. 
#These indicate the presence of relatively long contigs in all assemblies.
#The genome fraction percentages range from approximately 73.968% to 82.051%, 
#indicating that all assemblies captured a substantial portion of the reference genome 
#The duplication ratios are close to 1, suggesting a relatively low level of duplication in all assemblies.

#I chose the Spades assembly in all three files because the results were better

#spades-contigs-ERR204044_fasta has the highest genome fraction (77.022%)
#and a very low duplication ratio (1.003).

#spades-contigs-SRR15131330_fasta
#It has the highest genome fraction (82.051%) 
#and a relatively low duplication ratio (1.032).

# spades-contigs-SRR18214264_fasta
#It has a relatively high genome fraction (77.829%) 
#and a relatively low duplication ratio (1.028) 


#Mapping ERR204044

reads_dir="/home/bioinformatikai/HW2/outputs"
assembly_dir="/home/bioinformatikai/HW2/outputs/spades/ERR204044"
mapping_dir="/home/bioinformatikai/HW2/outputs/mapping"

mkdir -p ${mapping_dir}

read1="${reads_dir}/ERR204044_1_val_1.fq"
read2="${reads_dir}/ERR204044_2_val_2.fq"
assembly="${assembly_dir}/contigs.fasta"

bwa index ${assembly}
bwa mem -t 6 ${assembly} ${read1} ${read2} | samtools view -b -@ 6 - > ${mapping_dir}/ERR204044.bam
samtools sort -@ 6 ${mapping_dir}/ERR204044.bam -o ${mapping_dir}/ERR204044.sorted.bam
samtools index ${mapping_dir}/ERR204044.sorted.bam
samtools stats ${mapping_dir}/ERR204044.sorted.bam > ${mapping_dir}/ERR204044_map_stats.txt
bedtools genomecov -d -ibam /home/bioinformatikai/HW2/outputs/mapping/ERR204044.sorted.bam | awk '{c++}END{print c}'

#Mapping SRR15131330

reads_dir="/home/bioinformatikai/HW2/outputs"
assembly_dir="/home/bioinformatikai/HW2/outputs/spades/SRR15131330"
mapping_dir="/home/bioinformatikai/HW2/outputs/mapping"

mkdir -p ${mapping_dir}

read1="${reads_dir}/SRR15131330_1_val_1.fq"
read2="${reads_dir}/SRR15131330_2_val_2.fq"
assembly="${assembly_dir}/contigs.fasta"

bwa index ${assembly}
bwa mem -t 6 ${assembly} ${read1} ${read2} | samtools view -b -@ 6 - > ${mapping_dir}/SRR15131330.bam
samtools sort -@ 6 ${mapping_dir}/SRR15131330.bam -o ${mapping_dir}/SRR15131330.sorted.bam
samtools index ${mapping_dir}/SRR15131330.sorted.bam
samtools stats ${mapping_dir}/SRR15131330.sorted.bam > ${mapping_dir}/SRR15131330_map_stats.txt
bedtools genomecov -d -ibam /home/bioinformatikai/HW2/outputs/mapping/SRR15131330.sorted.bam | awk '{c++}END{print c}'

#Mapping SRR18214264

reads_dir="/home/bioinformatikai/HW2/outputs"
assembly_dir="/home/bioinformatikai/HW2/outputs/spades/SRR18214264"
mapping_dir="/home/bioinformatikai/HW2/outputs/mapping"

mkdir -p ${mapping_dir}

read1="${reads_dir}/SRR18214264_1_val_1.fq"
read2="${reads_dir}/SRR18214264_2_val_2.fq"
assembly="${assembly_dir}/contigs.fasta"

bwa index ${assembly}
bwa mem -t 6 ${assembly} ${read1} ${read2} | samtools view -b -@ 6 - > ${mapping_dir}/SRR18214264.bam
samtools sort -@ 6 ${mapping_dir}/SRR18214264.bam -o ${mapping_dir}/SRR18214264.sorted.bam
samtools index ${mapping_dir}/SRR18214264.sorted.bam
samtools stats ${mapping_dir}/SRR18214264.sorted.bam > ${mapping_dir}/SRR18214264_map_stats.txt
bedtools genomecov -d -ibam /home/bioinformatikai/HW2/outputs/mapping/SRR18214264.sorted.bam | awk '{c++}END{print c}'

#Evaluate mapping fraction as well as genome coverage from mapped reads

#ERR204044

#The number of properly paired reads is 5642658 out of 6051168, percentage of properly paired reads is approximately 92.9%.
#This indicates correct alignment and orientation.
#The average quality score (36.1) 
#The error rate is reported as 1.564794e-03
#Overall, these results indicate a successful mapping of the reads:
#The quality score and the percentage of properly paired reads are rather high
#And the error rate is low enough
#The genome coverage for the mapped reads in the sample ERR204044 is 2046030.

#SRR18214264

#the number of properly paired reads is 4185612 out of 4273832, percentage of properly paired reads is approximately 97.5%. 
#This indicates correct alignment and orientation
#The average quality score (32.0)
#The error rate is reported as: 9.022661e-03
#Overall, these results indicate a successful mapping of the reads:
#Eventhough the quality score is the lowest out of all it is still rather high
#The error rate is significamtly bigger than others
#However the percantage of properly paired reads is the highest out of all
#The genome coverage for the mapped reads in the sample SRR18214264 is 2036064

#SRR15131330

#the number of properly paired reads is 25524906 out of 28414526, percentage of properly paired reads is approximately 89.5%
#This indicates correct alignment and orientation
#he average quality is reported as 36.2
#The error rate is reported as: 1.895912e-03
#While the percentage of properly paired reads is the lowest out of all, it is still rather high
#The error rate a bit higher than in ERR204044 
#And the avarage quality is the highest out of all
#The genome coverage for the mapped reads in the sample SRR15131330 is 1943393


########################################################################################################################################

#GEPARD results

#ERR202044_SRR15131330_ragtag
#The diagonal (top left corner to bottom right corner) from the coordinate starting point shows 
#a direct correspondence and similarities between sequences. 
#We also see small lines and dots on one side or the other, which probably won't have much biological significance. 
#There are a few small gaps in the diagonal lines, indicating that the sequences in those positions do not correspond. 
#We can also observe a rectangle in the bottom right corner, which indicates that there were duplicates at the end of sequences.

#ERR204044_SRR18214264_RagTag
As before, there is a diagonal starting from the starting point, indicating similarities between sequences. 
#We also see small lines and dots on one side or the other. 
#In this case, the gaps in the diagonal lines are a bit bigger. 
#There is also a rectangle in the bottom right corner

#SRR15131330_SRR18214264_RagTag
#Here, the diagonal starts a bit to the right of the coordinate starting point. 
#There are even bigger gaps here, but overall the diagonal is rather seamless. 
#As before, there is a rectangle in the bottom right corner.

#BUSCO results:

#ERR204044:

#Complete BUSCOs: 398 out of 402 searched
#Complete and single-copy BUSCOs: 398 out of 402 searched
#Fragmented BUSCOs: 2
#Missing BUSCOs: 2
#The ERR204044 assembly shows a high level of completeness, 
#with the majority of BUSCOs being complete and single-copy. 
#There are only a few fragmented and missing BUSCOs, indicating good overall quality.

#SRR15131330:

#Complete BUSCOs: 396 out of 402 searched
#Complete and single-copy BUSCOs: 396 out of 402 searched
#Fragmented BUSCOs: 2
#Missing BUSCOs: 4
#he SRR15131330 assembly also demonstrates a high level of completeness, 
#with a similar number of complete and single-copy BUSCOs as the ERR204044 assembly. 
#However, there are a few more missing BUSCOs, indicating some gaps in the assembly

#SRR18214264:

#Complete BUSCOs: 398 out of 402 searched
#Complete and single-copy BUSCOs: 398 out of 402 searched
#Fragmented BUSCOs: 2
#Missing BUSCOs: 2
#The SRR18214264 assembly exhibits a completeness level comparable to the ERR204044 assembly, 
#with a high number of complete and single-copy BUSCOs. 
#There are only a few fragmented and missing BUSCOs, indicating good quality.


#Overall, the BUSCO results suggest that all three assemblies have a high degree of completeness, 
#with the majority of the expected BUSCOs being present as complete and single-copy genes. 
#The presence of fragmented and missing BUSCOs in some cases may indicate minor gaps or inaccuracies in the assemblies, 
#but the overall quality seems satisfactory.

makeblastdb -in /home/bioinformatikai/HW2/outputs/spades/ERR204044/contigs.fasta -dbtype prot -parse_seqids
tblastn -db /home/bioinformatikai/HW2/outputs/spades/ERR204044/contigs.fasta -query /home/bioinformatikai/HW2/CP015498/GCF_022832545.1_ASM2283254v1_protein.faa > /home/bioinformatikai/HW2/outputs/blast/blast_ERR204044.rez
tblastn -db /home/bioinformatikai/HW2/outputs/spades/ERR204044/contigs.fasta -query /home/bioinformatikai/HW2/CP015498/GCF_022832545.1_ASM2283254v1_protein.faa -out /home/bioinformatikai/HW2/outputs/blast/blast_ERR204044_tab.rez -outfmt 6
awk -F'\t' '{print $1}' blast_ERR204044_tab.rez | sort -u | wc -l


makeblastdb -in /home/bioinformatikai/HW2/outputs/spades/SRR15131330/contigs.fasta -dbtype prot -parse_seqids
tblastn -db /home/bioinformatikai/HW2/outputs/spades/SRR15131330/contigs.fasta -query /home/bioinformatikai/HW2/CP015498/GCF_022832545.1_ASM2283254v1_protein.faa > /home/bioinformatikai/HW2/outputs/blast/blast_SRR15131330.rez
tblastn -db /home/bioinformatikai/HW2/outputs/spades/SRR15131330/contigs.fasta -query /home/bioinformatikai/HW2/CP015498/GCF_022832545.1_ASM2283254v1_protein.faa -out /home/bioinformatikai/HW2/outputs/blast/blast_SRR15131330_tab.rez -outfmt 6
awk -F'\t' '{print $1}' blast_SRR15131330_tab.rez| sort -u | wc -l

makeblastdb -in /home/bioinformatikai/HW2/outputs/spades/SRR18214264/contigs.fasta -dbtype prot -parse_seqids
tblastn -db /home/bioinformatikai/HW2/outputs/spades/SRR18214264/contigs.fasta -query /home/bioinformatikai/HW2/CP015498/GCF_022832545.1_ASM2283254v1_protein.faa > /home/bioinformatikai/HW2/outputs/blast/blast_SSRR18214264.rez
tblastn -db /home/bioinformatikai/HW2/outputs/spades/SRR18214264/contigs.fasta -query /home/bioinformatikai/HW2/CP015498/GCF_022832545.1_ASM2283254v1_protein.faa -out /home/bioinformatikai/HW2/outputs/blast/blast_SSRR18214264_tab.rez -outfmt 6
awk -F'\t' '{print $1}' blast_SSRR18214264_tab.rez| sort -u | wc -l

makeblastdb -in /home/bioinformatikai/HW2/outputs/spades/ERR204044/contigs.fasta -dbtype nucl -parse_seqids
blastn -db /home/bioinformatikai/HW2/outputs/spades/ERR204044/contigs.fasta -query /home/bioinformatikai/HW2/CP015498/GCF_022832545.1_ASM2283254v1_genomic.fna > /home/bioinformatikai/HW2/outputs/blast/blast_nucl_ERR204044.rez
blastn -db /home/bioinformatikai/HW2/outputs/spades/ERR204044/contigs.fasta -query /home/bioinformatikai/HW2/CP015498/GCF_022832545.1_ASM2283254v1_genomic.fna -out /home/bioinformatikai/HW2/outputs/blast/blast_nucl_ERR204044_tab.rez -outfmt 6
awk -F'\t' '{print $1}' blast_nucl_ERR204044_tab.rez | sort -u | wc -l
awk '{ print $1 }' blast_nucl_ERR204044.rez | sort -u | wc -l

makeblastdb -in /home/bioinformatikai/HW2/outputs/spades/SRR15131330/contigs.fasta -dbtype nucl -parse_seqids
blastn -db /home/bioinformatikai/HW2/outputs/spades/SRR15131330/contigs.fasta -query /home/bioinformatikai/HW2/CP015498/GCF_022832545.1_ASM2283254v1_genomic.fna > /home/bioinformatikai/HW2/outputs/blast/blast_nucl_SRR15131330.rez
blastn -db /home/bioinformatikai/HW2/outputs/spades/SRR15131330/contigs.fasta -query /home/bioinformatikai/HW2/CP015498/GCF_022832545.1_ASM2283254v1_genomic.fna -out /home/bioinformatikai/HW2/outputs/blast/blast_nucl_SRR15131330_tab.rez -outfmt 6
awk -F'\t' '{print $1}' blast_nucl_SRR15131330_tab.rez | sort -u | wc -l
awk '{ print $1 }' blast_nucl_SRR15131330.rez | sort -u | wc -l

makeblastdb -in /home/bioinformatikai/HW2/outputs/spades/SRR18214264/contigs.fasta -dbtype nucl -parse_seqids
blastn -db /home/bioinformatikai/HW2/outputs/spades/SRR18214264/contigs.fasta -query /home/bioinformatikai/HW2/CP015498/GCF_022832545.1_ASM2283254v1_genomic.fna > /home/bioinformatikai/HW2/outputs/blast/blast_nucl_SRR18214264.rez
blastn -db /home/bioinformatikai/HW2/outputs/spades/SRR18214264/contigs.fasta -query /home/bioinformatikai/HW2/CP015498/GCF_022832545.1_ASM2283254v1_genomic.fna -out /home/bioinformatikai/HW2/outputs/blast/blast_nucl_SRR18214264_tab.rez -outfmt 6
awk -F'\t' '{print $1}' blast_nucl_SRR18214264_tab.rez | sort -u | wc -l
awk '{ print $1 }' blast_nucl_SRR18214264.rez | sort -u | wc -l


#Gene predictions
#ERR024044
#Rast= 2494
#Gene_Marks = 2311
#Blast nucl = 1873

#SRR15131330
#Rast = 2698
#Gene_Marks = 2550
#Blast nucl = 2380

#SRR18214264
#Rast = 2451
#Gene_Marks = 2338
#Blast nucl = 1500

#Overall RAST annotation generally predicted a higher number of genes 
#compared to GeneMark and BLAST predictions. With sequence 
#SRR15131330 the diference between gene predictions was least significant

